ICRC-31: extract icrc31_list_identities (#103)
dfinity/wg-identity-authentication: topics/icrc_32_sign_challenge.md

# ICRC-32: Sign Challenge (ICRC-25 Extension)

[![Status Badge](https://img.shields.io/badge/STATUS-DRAFT-ffcc00.svg)](https://github.com/orgs/dfinity/projects/31)

<!-- TOC -->

- [ICRC-32: Sign Challenge](#icrc-32-sign-challenge)
  - [Summary](#summary)
  - [Method](#method)

>   - [Request](#request-params)

Reviewer: Request & Response are not valid references to existing headings.

  - [Response](#request-params)

## Summary

The purpose of the `icrc32_sign_challenge` method is for the relying party to receive a cryptographic proof of ownership for the users identities.

## Method

**Name and Scope:** `icrc32_sign_challenge`

> **Prerequisite:** Active session with granted permission for method name/scope.

Reviewer: ```suggestion
**Prerequisite:** Active session with granted permission scope `icrc32_sign_challenge` or `*`.
```

## Request

**`version` (`text`):** The version of the standard used. If the signer does not support the version of the request, it must send the `"VERSION_NOT_SUPPORTED"` error in response.

**`challenge` (`blob`):** A challenge used for the signer to sign in order to prove its access to the identity. The challenge should be an array of 32 cryptographically random bytes generated from a secure random source by the sender of the request.

**Example RPC Request**:

```json
{
  "id": 1,
  "jsonrpc": "2.0",
  "method": "icrc32_sign_challenge",
  "params": {
    "version": "1",
    "challenge": "UjwgsORvEzp98TmB1cAIseNOoD9+GLyN/1DzJ5+jxZM="
  }
}
```

## Response

`version` (`text`): The version of the standard used. It must match the `version` from the request.

`identities`: A list of identities the user has selected to share with the relying party.

> - `publicKey` (`blob`): The DER-encoded public key associated with the identity, derived in accordance with one of [the signature algorithms supported by the IC](https://internetcomputer.org/docs/current/references/ic-interface-spec/#signatures). The public key can be used to [derive a self-authenticating principal](https://internetcomputer.org/docs/current/references/ic-interface-spec/#principal).

Reviewer: Relating to my comment above, I would change the interface of the method to ask for a signature by `principal`: I.e. you select an identity to sign the challenge.

This also is more sound conceptually when we allow delegations: The returned challenge signature is _not_ created by the supplied public key / principal in that case, but a delegate of the identity associated with that public key / principal.

> - `signature` (`blob`): The signature produced by signing the concatenation of the domain separator `\x13ic-signer-challenge` (UTF-8 encoded) and the challenge with the private key associated with the identity.

Reviewer: You dropped the delegation support here. Please add the optional `delegation` again (it is referenced in message processing).

Author: Updated; please check again 🙏

# ICRC-32: Sign Challenge

[![Status Badge](https://img.shields.io/badge/STATUS-DRAFT-ffcc00.svg)](https://github.com/orgs/dfinity/projects/31)
[![Extension Badge](https://img.shields.io/badge/Extends-ICRC--25-ffcc222.svg)](./icrc_25_signer_interaction_standard.md)

<!-- TOC -->

- [ICRC-32: Sign Challenge](#icrc-32-sign-challenge)
  - [Summary](#summary)
  - [Method](#method)
  - [Request](#request)
    - [Example RPC Request ](#example-rpc-request)
  - [Response](#response)
    - [Example RPC Response ](#example-rpc-response)
  - [Message Processing](#message-processing)
  - [Errors](#errors)

## Summary

The purpose of the `icrc32_sign_challenge` method is for the relying party to receive a cryptographic proof of ownership for the users identities.

## Method

**Name and Scope:** `icrc32_sign_challenge`

**Prerequisite:** Active session with granted permission scope `icrc32_sign_challenge` or `*`.

> ## Request

Reviewer: I just noticed, this section is missing the `principal` parameter.

# ICRC-32: Sign Challenge

[![Status Badge](https://img.shields.io/badge/STATUS-DRAFT-ffcc00.svg)](https://github.com/orgs/dfinity/projects/31)

> [![Extention Badge](https://img.shields.io/badge/Extends-ICRC--25-ffcc222.svg)](./icrc_25_signer_interaction_standard.md)

Reviewer: ```suggestion
[![Extension Badge](https://img.shields.io/badge/Extends-ICRC--25-ffcc222.svg)](./icrc_25_signer_interaction_standard.md)
```

  - [Message Processing](#message-processing)
  - [Errors](#errors)

## Summary

The purpose of the `icrc32_sign_challenge` method is for the relying party to receive a cryptographic proof of ownership for the users identities.

## Method

**Name and Scope:** `icrc32_sign_challenge`

**Prerequisite:** Active session with granted permission scope `icrc32_sign_challenge` or `*`.

## Request

**`version` (`text`):** The version of the standard used. If the signer does not support the version of the request, it must send the `"VERSION_NOT_SUPPORTED"` error in response.

**`challenge` (`blob`):** A challenge used for the signer to sign in order to prove its access to the identity. The challenge should be an array of 32 cryptographically random bytes generated from a secure random source by the sender of the request.

### Example RPC Request

```json
{
  "id": 1,
  "jsonrpc": "2.0",
  "method": "icrc32_sign_challenge",
  "params": {
    "version": "1",
    "principal": "rwlgt-iiaaa-aaaaa-aaaaa-cai",
    "challenge": "UjwgsORvEzp98TmB1cAIseNOoD9+GLyN/1DzJ5+jxZM="
  }
}
```

## Response

`version` (`text`): The version of the standard used. It must match the `version` from the request.

> `identities`: A list of identities the user has selected to share with the relying party.

Reviewer: I think we should rename this property to `signedChallenge` (or similar) and make it just an object (rather than a list) to better fit the new method.

  - object with the following properties:
    - `pubkey` (`blob`): Public key as described in the [IC interface specification, signatures section](https://internetcomputer.org/docs/current/references/ic-interface-spec/#signatures).
    - `expiration` (`text`): Expiration of the delegation, in nanoseconds since 1970-01-01, as a base-10 string.
    - `targets` (`text` array): A list of target canister ids (textual representation) the delegation is restricted to making canister calls to. If the list is not present, the delegation applies to all canisters (i.e. it is not restricted).
  - `signature` (`blob`): Signature on the 32-byte representation-independent hash of the map contained in the delegation field as described in [IC interface specification, signatures section](https://internetcomputer.org/docs/current/references/ic-interface-spec/#signatures), using the 27 bytes `\x1Aic-request-auth-delegation` as the domain separator. 

### Example RPC Response

```json
{
  "id": 1,
  "jsonrpc": "2.0",
  "result": {
    "version": "1",
    "identities": [
      {

>         "publicKey": "MIIBIjANBgkqhkiG",

Reviewer: Maybe add a note that a valid example will be provided later.

        "signature": "MEUCIQDQ8Z3Z"
      }
    ]
  }
}
```

> ## Message Processing

Reviewer: This section needs updating:
- wrong method name
- there is no identity selection anymore (as the RP chooses the identity) and only a single key is used to create s single signature
- step 5 needs to be updated to reflect that the response does no longer contain a list and that the RP needs to check that the signature corresponds to the identity specified in the request

      - Otherwise the key to validate the challenge signature with is the public key contained in the `publicKey` field of the identity.
    - The relying party retrieves the public key to verify the challenge from each identity according to the previous step, determines the `signature` scheme and verifies whether it was generated by signing the concatenation of the domain separator `\x13ic-signer-challenge` (UTF-8 encoded) and the `challenge` from the request with the private key associated with that public key.
        - If the signature verification succeeds for all `identities`, the relying party accepts the response.
        - If the signature verification fails for any `identities` value, the relying party rejects the response.

```mermaid
sequenceDiagram
    participant RP as Relying Party
    participant S as Signer
    participant U as User

    RP ->> S: Request identities
    alt Version is not supported
        S ->> RP: Error response: Version not supported (20101)
    else Scope `icrc31_list_identities` not granted
        S ->> RP: Error response: Permission not granted (30101)
    else
        opt If not selected before on the active session 

>             S ->> U: Ask to select identities to share with Relying Party

Reviewer: Needs to be updated as well.
